### DISCUSSION TITLE
코드를 실행 시켰을 때 출력이 어떻게 되는 지 말해주세요.

### DISCUSSION THREAD

**Discussion by publdaze** [publdaze/js-study#20] [변수]
- 코드를 실행 시켰을 때 출력이 어떻게 되는 지 말해주세요.
    ```js
    console.log(score);
    
    var score = 80;
    
    console.log(score);
    ```
    ```js
    console.log(score);
    
    score = 80;
    var score;
    
    console.log(score);
    ```
    ```js
    console.log(score);
    
    score = 80;
    var score = 50;
    
    console.log(score);
    ```
  - 호이스팅이란 무엇인가요?
    - 아래 코드의 경우는 어떻게 동작하나요?
      ```js
      console.log(score);
      
      let score = 80;
      ```

**Comment by publdaze** [comment]
> - 코드를 실행 시켰을 때 출력이 어떻게 되는 지 말해주세요.

```js
console.log(score); // undefined

var score = 80;

console.log(score); // 80
```
```js
console.log(score); // undefined

score = 80;
var score;

console.log(score); // 80
```
```js
console.log(score); // undefined

score = 80;
var score = 50;

console.log(score); // 50
```
>  - 호이스팅이란 무엇인가요?
- 자바스크립트에서 선언된 변수가 끌어올려지는 것 처럼 동작하는 걸 말함.
- var과 let, const 모두 호이스팅이 일어나지만 var은 선언과 동시에 초기화가 일어나서 값 할당 전 reference에러가 발생하지 않지만, let과 const는 선언과 초기화가 분리되어 일어나서 선언과 초기화 사이인 일시적 사각 지대에서는 reference에러가 발생함.

>    - 아래 코드의 경우는 어떻게 동작하나요?

```js
console.log(score);

let score = 80;
```
- reference에러가 발생

**Comment by pipisebastian** [comment]
>코드를 실행 시켰을 때 출력이 어떻게 되는 지 말해주세요.

1. undefined, 80
2. undefined, 80
3. undefined, 50

> 호이스팅이란 무엇인가요?

변수의 선언 단계 는 런타임 이전에 실행되며, 실행 컨텍스트 안에 식별자 정보(key - 변수 이름, value - 변수 값)가 저장됩니다.
 이는 변수 선언이 맨 처음 실행되어, 마치 코드 최상단에 위치한 것처럼 보이기 때문에, 호이스팅이라고도 합니다.

>아래 코드의 경우는 어떻게 동작하나요

1. reference Error
해당 코드의 경우, 마치 호이스팅이 발생하지 않는 것처럼 보이는데요. 이는 let,const의 경우 var와 다르게 선언과 초기화단계 사이에 텀이 존재하기 때문입니다. 해당 키워드들은 초기화단계가 런타임에서 코드를 한줄한줄 실행하며, 변수 선언문에 도달했을때 undefined로 초기화가 됩니다. 
그렇기 때문에 그 사이에 접근할경우, 초기화가 되지 않은 상태이기 때문에 reference error가 발생합니다.
